zcash_client_backend/src/proto.rs

```diff
@@ -142,3 +164,226 @@ impl compact_formats::CompactSaplingSpend {
+            .iter()
+            .map(|rec_note| proposal::ProposedInput {
+                txid: rec_note.txid().as_ref().to_vec(),
+                pool: 2,
+                index: rec_note.output_index().into(),
+                value: rec_note.value().into(),
+            })
+            .chain(
+                value
+                    .transparent_inputs()
+                    .iter()
+                    .map(|utxo| proposal::ProposedInput {
+                        txid: utxo.outpoint().hash().to_vec(),
+                        pool: 0,
+                        index: utxo.outpoint().n(),
+                        value: utxo.txout().value.into(),
+                    }),
+            )
+            .collect();
+
+        let balance = Some(proposal::TransactionBalance {
+            proposed_change: value
+                .balance()
+                .proposed_change()
+                .iter()
+                .map(|cv| match cv {
+                    ChangeValue::Sapling { value, memo } => proposal::ChangeValue {
+                        value: Some(proposal::change_value::Value::SaplingValue(
+                            proposal::SaplingChange {
+                                amount: value.into(),
+                                memo: memo.as_ref().map(|memo_bytes| proposal::MemoBytes {
+                                    value: memo_bytes.as_slice().to_vec(),
+                                }),
+                            },
+                        )),
+                    },
+                })
+                .collect(),
+            fee_required: value.balance().fee_required().into(),
+        });
+
+        #[allow(deprecated)]
+        Some(proposal::Proposal {
+            proto_version: PROPOSAL_SER_V1,
+            transaction_request,
+            inputs,
+            balance,
+            fee_rule: match value.fee_rule() {
+                StandardFeeRule::PreZip313 => proposal::FeeRule::PreZip313,
+                StandardFeeRule::Zip313 => proposal::FeeRule::Zip313,
+                StandardFeeRule::Zip317 => proposal::FeeRule::Zip317,
+            }
+            .into(),
+            min_confirmations: value.min_confirmations().into(),
+            min_target_height: value.min_target_height().into(),
+            min_anchor_height: value.min_anchor_height().into(),
+            is_shielding: value.is_shielding(),
+        })
+    }
+
+    /// Attempts to parse a [`Proposal`] based upon a supported [`StandardFeeRule`] from its
+    /// protobuf representation.
+    pub fn try_into_standard_proposal<P: consensus::Parameters, DbT: WalletRead>(
+        &self,
+        params: &P,
+        wallet_db: &DbT,
+    ) -> Result<Proposal<StandardFeeRule, DbT::NoteRef>, ProposalError<DbT::Error>> {
+        match self.proto_version {
+            PROPOSAL_SER_V1 => {
+                let transaction_request =
+                    TransactionRequest::from_uri(params, &self.transaction_request)?;
```

Comment:
At present there's not any integrity checking, because we don't keep track of proposals that have been returned to the caller. That seems like a reasonable feature to add; we could add an integrity check and store the proposal or at least its hash, but I think that should be done in a separate PR to this one. If that's a feature that you want, please add an issue for it and we can discuss designs.

Reply:
I think I was not specific, I was referring to this comment https://github.com/zcash/librustzcash/pull/891/files#r1301913308, where the proposal protobuf can be changed and the Request representation can be out of sync with the protobuf itself.

Reply:
This is a good catch; parsing the `Proposal` should do balance integrity checks.

Reply:
Fixed in a5d2a61624873161032e58fc4b5b44d2f46eea4e